src/main/java/org/magicdgs/readtools/utils/mappability/gem/GemMappabilityReader.java

```diff
@@ -0,0 +1,274 @@
+    private static final String APPROX_THRESHOLD_HEADER_TEXT = "APPROXIMATION THRESHOLD";
+    private static final String MAX_MISMATCH_HEADER_TEXT = "MAX MISMATCHES";
+    private static final String MAX_ERRORS_HEADER_TEXT = "MAX ERRORS";
+    private static final String MAX_INDEL_LENGTH_HEADER_TEXT = "MAX BIG INDEL LENGTH";
+    private static final String MIN_MATCH_HEADER_TEXT = "MIN MATCHED BASES";
+    private static final String STRATA_AFTER_BEST_HEADER_TEXT = "STRATA AFTER BEST";
+    private static final String ENCODING_HEADER_TEXT = "ENCODING";
+
+    // pattern for the encoding (char to value range)
+    private static final Pattern ENCODING_PATTERN = Pattern.compile("'(.)'~\\[(\\d+)-(\\d+)]");
+
+    // stored for error message and get the identity of the path
+    private final Path path;
+    // header from the file, read on construction
+    private final GemMappabilityHeader header;
+
+    private FastLineReader reader;
+
+    // iteration values to keep track of the current sequence
+    private String currentSequence = null;
+    private int currentSequencePosition = -1;
+
+    // logger
+    private final Logger logger = LogManager.getLogger(this.getClass());
+
+    /**
+     * Loads a gem-mappability formatted file.
+     *
+     * @param path path where the file is located.
+     *
+     * @throws IOException if the reader cannot be open.
+     */
+    public GemMappabilityReader(final Path path) throws IOException {
+        this.path = Utils.nonNull(path);
+        // TODO: accept compressed inputs (https://github.com/magicDGS/ReadTools/issues/482)
+        this.reader = new FastLineReader(Files.newInputStream(this.path));
+        this.header = readHeader();
+    }
+
+    /**
+     * Testing constructor.
+     *
+     * @param reader reader starting from the encodings.
+     * @param header header with the meta-data storage.
+     */
+    @VisibleForTesting
+    protected GemMappabilityReader(final FastLineReader reader, final GemMappabilityHeader header) {
+        this.reader = reader;
+        this.path = null;
+        this.header = header;
+    }
+
+    /**
+     * Gets the header associated with this file.
+     *
+     * @return gem-mappability header.
+     */
+    public GemMappabilityHeader getHeader() {
+        return header;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (reader == null) {
+            return false;
+        }
+        // first skip the new lines to ensure that hasNext is accurate
+        reader.skipNewlines();
+        return !reader.eof();
+    }
+
+    @Override
+    public GemMappabilityRecord next() {
+        // throw if there is no next
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        // changing to next sequence
+        if (isAtHeaderLine()) {
+            this.currentSequence = readLine().substring(1);
+            this.currentSequencePosition = 0;
```

Comment:
Better change here for 1 and in line 167 for `this.currentSequencePosition++` - otherwise the exception in the range check will have invalid coordinates

Reply:
Done.